### PR TITLE
Length Service Input

### PR DESCRIPTION
**Pull request by CaitW**
Adds a Length service input type per issue #9 

* Does the select input need an ID field? What do we use the ID field for?

### CODE REVIEW

File: src/gm3/util.js [hunk 1]
```diff
@@ -436,3 +436,21 @@ export function metersAreaToUnits(meters, units) {
...
+        'ft': 0.3048,
+        'yd': 0.9144,
+        'mi': 1609.347,
+        'in': 0.0254,
+        'm': 1,
+        'km': 1000,
+        "ch": 20.11684
+    };
+    let convertedLength = length * equivalentMeters[srcUnits] / equivalentMeters[destUnits];
+    if (typeof convertedLength === "number" && convertedLength >= 0) {
```
Comment (theduckylittle) [inline src/gm3/util.js]: This check should be removed.  Negative distances are a real thing and returning NaN is also valid and should be left up to the calling function to determine how to best handle that.

**Comment by theduckylittle** [comment]
@CaitW Please checkout my comment, everything else looks good,.

**Comment by klassenjs** [comment]
Why are we duplicating the conversion table in convertLength and
metersLengthToUnits?

Is the hash lookup or the switch faster?

**Comment by theduckylittle** [comment]
That's a cleanup step we need to make.  I did the previous "metersLengthToUnits" function as a quicky  for making the measure tool work properly.  Ostensibly, metersLengthToUnits should be retired in favor of the new function.  But that's out of scope of the issue.  I can write up an issue or just mentally note that after this merge is done the measure tool should be converted.  

Also the switch vs hash is an interesting question.  To my understanding the switch is probably faster and more memory efficient but the code in convertUnits is fairly elegant.  Putting the hash in a more permanent memory area (as a const ahead of the function) would probably increase the speed as there wouldn't be the need to reinitialize the object with every function call.  (Lazy dev coming... )Of course, the JIT/VM may just do that automatically and not really be an issue. (end)

**Comment by klassenjs** [comment]
Then I would recommend the documentation of the constants be included in
the new function.

On May 5, 2017 08:46, "theduckylittle" <notifications@github.com> wrote:

That's a cleanup step we need to make. I did the previous
"metersLengthToUnits" function as a quicky for making the measure tool work
properly. Ostensibly, metersLengthToUnits should be retired in favor of the
new function. But that's out of scope of the issue. I can write up an issue
or just mentally note that after this merge is done the measure tool should
be converted.

Also the switch vs hash is an interesting question. To my understanding the
switch is probably faster and more memory efficient but the code in
convertUnits is fairly elegant. Putting the hash in a more permanent memory
area (as a const ahead of the function) would probably increase the speed
as there wouldn't be the need to reinitialize the object with every
function call. (Lazy dev coming... )Of course, the JIT/VM may just do that
automatically and not really be an issue. (end)

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/geomoose/gm3/pull/30#issuecomment-299468803>, or mute
the thread

**Comment by theduckylittle** [comment]
re-reading the PR code post-coffee -- there really should be more comments than just the license header.

Please ensure that each class has a description of what it does and each function which is non-standard (IE, not render, onChange, etc.) is given a proper description, parameter values, and expect return value description.

File: src/gm3/components/serviceInputs/length.jsx [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+/**
+ * Service input type for specifying length
+ * - Renders a numeric text input and a select-list of units
+ * - User-specified length is converted to its equivalent in meters and stored
+ *   in the component state
+ *
+ * @extends TextInput
+ */
+import React from 'react';
```
Comment (theduckylittle) [inline src/gm3/components/serviceInputs/length.jsx]: New line

File: src/gm3/components/serviceInputs/length.jsx [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+import TextInput from './text.jsx';
+import { convertLength } from '../../util.js';
+export default class LengthInput extends TextInput {
```
Comment (theduckylittle) [inline src/gm3/components/serviceInputs/length.jsx]: new line

File: src/gm3/components/serviceInputs/length.jsx [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+            value: 'km'
+        }, {
+            label: 'Chains',
+            value: 'ch'
+        }];
+        this.state = {
+            value: props.field.default ? props.field.default : 0
+        };
+    }
+    onChange() {
```
Comment (theduckylittle) [inline src/gm3/components/serviceInputs/length.jsx]: new line

File: src/gm3/components/serviceInputs/length.jsx [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        let lengthInput = parseFloat(this.lengthInput.value);
+        let unitInput = this.unitInput.value;
+        let lengthInMeters = convertLength(lengthInput, unitInput, 'm');
+        this.setState({ value: lengthInMeters });
+        this.setValue(this.getName(), lengthInMeters);
+    }
+    /**
```
Comment (theduckylittle) [inline src/gm3/components/serviceInputs/length.jsx]: new line

File: src/gm3/util.js [hunk 6]
```diff
@@ -436,3 +436,23 @@ export function metersAreaToUnits(meters, units) {
             return meters;
     }
 }
+/** Converts numeric lengths between given units
```
Comment (theduckylittle) [inline src/gm3/util.js]: new line

File: src/gm3/util.js [hunk 7]
```diff
@@ -436,3 +436,23 @@ export function metersAreaToUnits(meters, units) {
...
+ *
+ * @param {number} length - Length
+ * @param {string} srcUnits - Source unit
+ * @param {string} destUnits - Destination unit
+ * @return {number} Converted length
+ */
+export function convertLength(length, srcUnits, destUnits) {
+    // US survey feet, miles
+    const equivalentMeters = {
```
Comment (theduckylittle) [inline src/gm3/util.js]: As a small optimization (and to make the constants available to others, this should get moved OUT of the function and put into it's own declaration.  E.G. "const EQUIVALENT_METERS = { ..}" on line 439.

**Comment by theduckylittle** [comment]
"Don't fear the void".  I added comments suggesting newline to help break up the visual flow the code a little bit more.  Not having the blank lines after function declarations can make the code look denser than it really needs to be and harder to visually breakup by others.

**Pull request merged**